### PR TITLE
Add build stage to Dockerfile

### PR DESCRIPTION
**Pull request by pjonsson**
The image currently has 840 security
vulnerabilities according to Trivy.
Many of those vulnerabilities are
in the development packages, so
add a build stage to the Dockerfile
so the development packages do not
end up in the final image. This reduces
the final image size by roughly
1/3 of its size.

Since everything is being changed,
also replace wget with curl, so we
get error messages on HTTP failures
(wget -q silences everything including
error printouts, and the behavior
cannot be overridden).

### CODE REVIEW

**Comment by pjonsson** [comment]
@alexgleith according to the commit logs it was you who removed the previously existing build-stage of the Dockerfile. What do you say about re-introducing it? There's a substantial saving in image size, and a fair amount of fewer security vulnerabilities flagged by Trivy.

This PR builds the binary wheels in the build stage of the container and installs them in the "real" container. I'm not sure how to put the wheels in the requirements.txt so it works with pip-compile, or if we perhaps could drop the pip-compile step completely (which would reduce the number of releases required since a rebuild of the image would give security updates).

And as a general remark, I cannot understand how the CI tests did not trigger the eodatasets3 distutils dependency problem until now.

File: index/Dockerfile [hunk 1]
```diff
@@ -37,26 +68,30 @@ RUN apt-get update \
 ENV VIRTUAL_ENV=/virtualenv/python3.12 \
```
Comment (alexgleith) [inline index/Dockerfile:68]: You could set this earlier, and use it in the previous step.

**Review by alexgleith — COMMENTED** [review]
Looks good to me. Does it reduce the image size significantly?

The reason I like single-stage builds is that it keep things simpler... but this looks neat and tidy, and with Python 12 requiring virtual environments, it's a good way to do it.

One idea, and it's kinda complex too, is to use `uv` to do the installing... it's super fast. Does introduce a new change, though. Example here for doing that: https://github.com/auspatious/ldn-land-productivity/blob/main/Dockerfile

File: index/Dockerfile [hunk 2]
```diff
@@ -28,35 +30,64 @@ RUN apt-get update \
...
+ENV LC_ALL=C.UTF-8 \
+    LANG=C.UTF-8 \
+    PATH=/app/bin:$PATH \
+    PYTHONDONTWRITEBYTECODE=1 \
+    PYTHONUNBUFFERED=1
+
+RUN export DEBIAN_FRONTEND=noninteractive \
+    && apt-get update \
+    && apt-get upgrade -y \
+    # Developer convenience
```
Comment (alexgleith) [inline index/Dockerfile:65]: This comment isnt required now

File: index/Dockerfile [hunk 3]
```diff
@@ -28,35 +30,64 @@ RUN apt-get update \
...
+
+COPY --from=builder --link --chown=ubuntu:ubuntu /app /app
+COPY --from=builder --link /build/*.txt /conf/
 
 # Copy Datacube bootstrapping and other scripts
-ADD ./assets /code
-RUN wget -q https://github.com/opendatacube/datacube-dataset-config/archive/refs/heads/main.zip \
-    -O /tmp/datacube-dataset-config.zip \
+COPY --link ./assets /code
+RUN cat /conf/version.txt \
```
Comment (alexgleith) [inline index/Dockerfile:84]: Maybe cat the version last?

**Review by alexgleith — DISMISSED** [review]
Looks good! Just two minor comments

File: index/Dockerfile (hunk 2 above)
Comment (pjonsson) [reply index/Dockerfile]: Fixed.

File: index/Dockerfile (hunk 3 above)
Comment (pjonsson) [reply index/Dockerfile]: Yeah, makes sense, fixed.

**Review by alexgleith — APPROVED** [review]

**Comment by pjonsson** [comment]
Seems the build failed for the job that pushes the image. I have no idea what is causing it, but I filed bug report on the action (https://github.com/whoan/docker-build-with-cache-action/issues/161) and made a PR that adds Dependabot for Github actions/Dockerfile updates (https://github.com/whoan/docker-build-with-cache-action/pull/162) in case it is related to the Docker version the action uses.

**Pull request merged**